returnn/tf/util/data.py

```diff
@@ -295,6 +295,23 @@ def get_uniq_collection(cls, tags, is_equal_opts=None):
         res.append(tag)
     return res
 
+  def get_unbroadcast_dim(self):
+    """
+    Infers the dim this axis should have if unbroadcasted.
+    If `self.src_data` has a placeholder, will use the shape from there.
+    Otherwise, uses `self.dimension` (if static) or `self.dyn_size` (if dynamic).
+    :rtype: int|Tensor
+    """
+    if self.src_data is not None and self.src_axis is not None and self.src_data.placeholder is not None:
+      return tf.shape(self.src_data.placeholder)[self.src_axis]
+    if self.dimension is not None:
+      return self.dimension
+    if self.dyn_size is not None:
+      return tf.math.reduce_max(self.dyn_size)
+    if self.kind == self.Types.Batch:
```

Comment:
This should be the second check. `LayerBase.get_recent_layer().get_batch_dim()` should be preferred over `tf.shape(self.src_data.placeholder)[self.src_axis]`. `get_batch_dim` will just depend on the input data (extern data), nothing else, so is immediately available, and so TF doesn't need to wait until `self.src_data.placeholder` is available.